sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala

```diff
@@ -164,4 +167,45 @@ class PartitionedWriteSuite extends QueryTest with SharedSparkSession {
       assert(e.getMessage.contains("Found duplicate column(s) b, b: `b`;"))
     }
   }
+
+  test("SPARK-27194 SPARK-29302: Fix commit collision in dynamic partition overwrite mode") {
+    withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key ->
+      SQLConf.PartitionOverwriteMode.DYNAMIC.toString,
+      SQLConf.FILE_COMMIT_PROTOCOL_CLASS.key ->
+        classOf[ConstantJobIdCommitProtocol].getName) {
+      withTempDir { d =>
+        withTable("t") {
+          sql(
+            s"""
+               | create table t(c1 int, p1 int) using parquet partitioned by (p1)
+               | location '${d.getAbsolutePath}'
+            """.stripMargin)
+
+          // File commit protocol is ConstantJobIdCommitProtocol, whose jobId is 'jobId'.
+          val stagingDir = new File(d, ".spark-staging-jobId")
+          stagingDir.mkdirs()
+          val conflictTaskFile = new File(stagingDir, "part-00000-jobId-c000.snappy.parquet")
+          conflictTaskFile.createNewFile()
```

Comment:
Sorry, the ut is wrong, I have fix it.
          val stagingDir = new File(d, ".spark-staging-jobId")
          stagingDir.mkdirs()
          val stagingPartDir = new File(stagingDir, "p1=2")
          stagingPartDir.mkdirs()
          val conflictTaskFile = new File(stagingPartDir, "part-00000-jobId.c000.snappy.parquet")
          conflictTaskFile.createNewFile()


I also recreated another PR, https://github.com/apache/spark/pull/28989,
In this Pr, I define a Spark staging output committer to leverage OutputCommitCoordinator

Reply:
Thank you for reminding, this UT is based on your pr #26339 , I'll correct it

Reply:
So, @turboFei you would prefer #28989 ? If so, maybe a bit clarification in this pr thread should be added.

Reply:
@advancedxy thanks